language/bytecode-verifier/src/type_safety.rs

```diff
@@ -741,3 +850,52 @@ fn instantiate(token: &SignatureToken, subst: &Signature) -> SignatureToken {
         }
     }
 }
+
+// Checks if the given types are permissible as struct instantiation for a vector element type
+// NOTE: references are not permitted.
+fn check_vector_element_sig_tokens(tys: &[SignatureToken]) -> PartialVMResult<()> {
+    for ty in tys {
+        check_vector_element_sig_token(ty)?;
+    }
+    Ok(())
+}
+
+// Checks if the given type is permissible as vector element type.
+// NOTE: reference types are not permitted.
+fn check_vector_element_sig_token(ty: &SignatureToken) -> PartialVMResult<()> {
+    use SignatureToken::*;
+    match ty {
+        U8 | U64 | U128 | Bool | Address | Signer | Struct(_) | TypeParameter(_) => Ok(()),
+        Reference(_) | MutableReference(_) => {
+            Err(PartialVMError::new(StatusCode::INVALID_VECTOR_ELEMENT_TYPE)
+                .with_message("reference not allowed as vector element type".to_string()))
+        }
+        Vector(ty) => check_vector_element_sig_token(ty),
+        StructInstantiation(_, type_arguments) => check_vector_element_sig_tokens(type_arguments),
+    }
+}
```

Comment:
Should be done in instruction concistency, or signatures.rs

Reply:
Check relocated to `signature.rs`, and hopefully, using the ability checking in the correct way.